common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+  // if passed. The default permissions in the `creation_mode` are `0700` which
+  // is more suitable for directories. There are no extra flags that can be
+  // passed.
+  //
+  // As with other open routines, when creating a directory, only the leaf
+  // component can be created by the call to this routine.
+  //
+  // When creating a directory with `CreateNew`, this routine works to be safe
+  // even in the presence of adversarial, concurrent operations. Specifically,
+  // even if the current (parent) directory allows different users to create
+  // subdirectories, and an adversarial user attempts to detect a new
+  // subdirectory created by calling this routine and replace it with one under
+  // their control, this routine reliably will return the directory under this
+  // process's control or produce an error due to an existing directory.
+  // However, no validation is done on any prefix path components leading to the
+  // leaf component created. When securely creating directories, the initial
+  // creation should typically have a single component from a known and opened
+  // parent directory.
+  auto OpenDir(std::filesystem::path path,
+               CreationFlags creation_flags = OpenExisting,
+               ModeType creation_mode = 0700) -> ErrorOr<Dir, PathError>;
+
+  // Reads the file at the provided path to a string.
+  //
+  // This is a convenience wrapper for opening the path, reading the returned
+  // file to a string, and closing it. Errors from any step are returned.
+  auto ReadFileToString(std::filesystem::path path)
+      -> ErrorOr<std::string, PathError>;
+
+  // Writes the provided `content` to the provided path.
+  //
+  // This is a convenience wrapper for opening the path, creating it according
+  // to `creation_flags` as necessary, writing `content` to it, and closing it.
+  // Errors from any step are returned.
+  auto WriteFileFromString(std::filesystem::path path, llvm::StringRef content,
+                           CreationFlags creation_flags = CreateAlways)
+      -> ErrorOr<Success, PathError>;
+
+  // Changes the current working directory to this directory.
+  auto Chdir() -> ErrorOr<Success, FdError>;
+
+  // Changes the current working directory to the provided path.
+  //
+  // An error if the provided path is not a directory. Does not open the
+  // provided path as a directory, but it will be available as the current
+  // working directory via `Cwd()`.
+  auto Chdir(std::filesystem::path path) -> ErrorOr<Success, PathError>;
+
+  // Creates a symlink at the provided path with the contents of `target`.
+  //
+  // Note that the target of a symlink is an arbitrary string and there is no
+  // error checking on whether it exists or is sensible. Also, the target string
+  // set will be up to the first null byte in `target`, regardless of its
+  // `size`. This will not overwrite an existing symlink at the provided path.
+  auto Symlink(std::filesystem::path path, const std::string& target)
```

Comment:
```suggestion
  auto Symlink(std::filesystem::path path, llvm::StringRef target)
```
If we don't use `path`, we should not force use of an owning `string`

Reply:
Ah, I see, you want a nul terminator in the body. WTB CStringRef.